src/Java.Interop.Tools.JavaTypeSystem/JavaModels/JavaTypeModel.cs

```diff
@@ -0,0 +1,121 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Java.Interop.Tools.JavaTypeSystem.Models
+{
+	public abstract class JavaTypeModel : IJavaResolvable
+	{
+		/// <summary>
+		/// Only the type's name, does not include parent type name for nested type.
+		/// </summary>
+		public string Name { get; }
+
+		/// <summary>
+		/// Includes parent type name(s) if type is nested (period separator). ex: Manifest.permission
+		/// </summary>
+		public string NestedName { get; set; }
+
+		public string Visibility { get; }
+		public bool IsAbstract { get; }
+		public bool IsFinal { get; }
+		public string Deprecated { get; }
+		public bool IsStatic { get; }
+		public string ExtendedJniSignature { get; }
+		public bool IsReferenceOnly { get; internal set; }
+
+		public JavaPackage Package { get; }
+		public JavaTypeModel? ParentType { get; internal set; }
+		public List<JavaTypeModel> NestedTypes { get; } = new List<JavaTypeModel> ();
+
+		public JavaTypeParameters TypeParameters { get; }
+		public List<JavaImplementsModel> Implements { get; } = new List<JavaImplementsModel> ();
+		public List<JavaTypeReference> ImplementsModels { get; } = new List<JavaTypeReference> ();
+		public List<JavaMethodModel> Methods { get; } = new List<JavaMethodModel> ();
+		public List<JavaFieldModel> Fields { get; } = new List<JavaFieldModel> ();
+
+		public Dictionary<string, string> PropertyBag { get; } = new Dictionary<string, string> ();
+
+		protected JavaTypeModel (JavaPackage javaPackage, string javaNestedName, string javaVisibility, bool javaAbstract, bool javaFinal, string deprecated, bool javaStatic, string jniSignature)
+		{
+			Package = javaPackage;
+			NestedName = javaNestedName.Replace ('$', '.');
+			Name = NestedName.LastSubset ('.');
+			Visibility = javaVisibility;
+			IsAbstract = javaAbstract;
+			IsFinal = javaFinal;
+			Deprecated = deprecated;
+			IsStatic = javaStatic;
+			ExtendedJniSignature = jniSignature;
+
+			TypeParameters = new JavaTypeParameters (this);
+		}
+
+		/// <summary>
+		/// Returns string containing package name, parent name, and type's name. (ex: 'java.util.ArrayList.Keys')
+		/// </summary>
+		public string FullName {
```

Comment:
Possibly a "crazy" idea, but could we *avoid* "FullName" as a construct? It's not immediately obvious at callsites if this is a *Java* full name or a C# full name.

Even the `<summary/>` docs are inconsistent with the code: docs suggest a Java name ("ex `java.util.ArrayList.Keys`"), while code suggests a C# name (replaces `boolean` with `bool`).

Furthermore, in the context of C#, it's possible for the same Java type to be bound by multiple different assemblies, which would necessitate "AssemblyQualifiedNames"…which can't be used in generated C# source code.

Reply:
I agree that `FullName` can get problematic in `generator`.

I tried to keep this context a little clearer as this library (`JavaTypeSystem`) only deals in Java types, and the class is called `JavaTypeModel`.  Originally every property was prefixed with "Java" like `JavaFullName/JavaVisibility/JavaStatic/etc`, but it seemed redundant in a `JavaTypeModel` class.

The `bool` thing appears to be leftover code from some earlier stuff.  We still output `boolean`, so I have removed this unintended C#-ism to keep this library pure "Java".

Reply:
> I tried to keep this context a little clearer as this library (`JavaTypeSystem`) only deals in Java types

which may address my earlier questions/suggestions around naming, e.g. to prefer System.Reflection-style names for similar purposes, such as `DeclaringType`, `FieldType`, etc.

The benefit to System.Reflection-style names is that, I assume, more C# developers will be familiar with those nouns, and thus the resulting code will be easier for other people to understand.

The alternative to using System.Reflection-style names is to use "Java-style" names, of which there are two: Java language as per [JLS](https://docs.oracle.com/javase/specs/jls/se16/html/index.html), or the [JVM spec](https://docs.oracle.com/javase/specs/jvms/se7/html/index.html), assuming that those use different names.  (One hopes that they're consistent with each other, but in the off chance they're not…)

So what nouns do the JLS & JVM use? JLS uses "declared" and synonyms: https://docs.oracle.com/javase/specs/jls/se16/html/jls-8.html#jls-8.3

> The notion of overriding includes methods that override another from some subclass of their declaring class.

"Parent" doesn't appear anywhere in jls-8.

JVM is similar, e.g. https://docs.oracle.com/javase/specs/jvms/se7/html/jvms-4.html#jvms-4.6: no mention of "parent", many mentions of "declared", "declaring", "declares", etc.

Reply:
Switched to `Delcared*`.